This also now fixes #1428 by specifying utf-8 as the default output encoding. Is this a good idea, @luizirber? All the tests pass with only one rather simple fix needed, so... 🤷 

A remaining nit in this PR is a Rust panic triggered in the `sourmash compute` code -- it appears that asking for an md5sum representation of a signature containing multiple MinHashes is not handled in the rust code. @luizirber is there any possible fix for this or should I abandon this functionality for output signatures for now?

error triggered by test `test_do_sourmash_sketchdna_output_zipfile`:

```
Traceback (most recent call last):
  File "/Users/t/dev/sourmash/tests/sourmash_tst_utils.py", line 105, in runscript  
    status = _runscript(scriptname)
  File "/Users/t/dev/sourmash/tests/sourmash_tst_utils.py", line 59, in _runscript  
    namespace)
  File "/Users/t/miniconda3/envs/py37/bin/sourmash", line 33, in <module>
    sys.exit(load_entry_point('sourmash', 'console_scripts', 'sourmash')())
  File "/Users/t/dev/sourmash/src/sourmash/__main__.py", line 13, in main
    return mainmethod(args)
  File "/Users/t/dev/sourmash/src/sourmash/cli/sketch/dna.py", line 89, in main
    return sourmash.command_sketch.dna(args)
  File "/Users/t/dev/sourmash/src/sourmash/command_sketch.py", line 216, in dna
    _execute_sketch(args, signatures_factory)
  File "/Users/t/dev/sourmash/src/sourmash/command_sketch.py", line 196, in _execute_sketch
    _compute_individual(args, signatures_factory)
  File "/Users/t/dev/sourmash/src/sourmash/command_compute.py", line 214, in _compute_individual
    save_siglist(siglist, args.output)
  File "/Users/t/dev/sourmash/src/sourmash/command_compute.py", line 273, in save_siglist
    save_sig.add(ss)
  File "/Users/t/dev/sourmash/src/sourmash/sourmash_args.py", line 670, in add
    md5 = ss.md5sum()
  File "/Users/t/dev/sourmash/src/sourmash/signature.py", line 82, in md5sum
    return decode_str(self.minhash._methodcall(lib.kmerminhash_md5sum))
  File "/Users/t/dev/sourmash/src/sourmash/signature.py", line 46, in minhash
    self._methodcall(lib.signature_first_mh)
  File "/Users/t/dev/sourmash/src/sourmash/utils.py", line 25, in _methodcall
    return rustcall(func, self._get_objptr(), *args)
  File "/Users/t/dev/sourmash/src/sourmash/utils.py", line 78, in rustcall
    raise exc
sourmash.exceptions.Panic: sourmash panicked: thread 'unnamed' panicked with 'not implemented' at src/core/src/ffi/signature.rs:172
```

--- a/doc/command-line.md
+++ b/doc/command-line.md
@@ -886,7 +886,7 @@ This behavior is triggered by the requested output filename --
 * to save to JSON signature files, use `.sig`; `-` will send JSON to stdout.
 * to save to gzipped JSON signature files, use `.sig.gz`;
 * to save to a Zip file collection, use `.zip`;
-* to save to a directory, use a name ending in `/`; the directory will be created if it doesn't exist;
+* to save signature files to a directory, use a name ending in `/`; the directory will be created if it doesn't exist;
 
 All of these save formats can be loaded by sourmash commands, too.
 

--- a/src/sourmash/command_compute.py
+++ b/src/sourmash/command_compute.py
@@ -267,10 +267,23 @@ def set_sig_name(sigs, filename, name=None):
 
 
 def save_siglist(siglist, sigfile_name):
+    import sourmash
+
     # save!
     with sourmash_args.SaveSignaturesToLocation(sigfile_name) as save_sig:
         for ss in siglist:
-            save_sig.add(ss)
+            try:
+                save_sig.add(ss)
+            except sourmash.exceptions.Panic:
+                # this deals with a disconnect between the way Rust
+                # and Python handle signatures; Python expects one
+                # minhash (and hence one md5sum) per signature, while
+                # Rust supports multiple. For now, go through serializing
+                # and deserializing the signature! See issue #1167 for more.
+                json_str = sourmash.save_signatures([ss])
+                for ss in sourmash.load_signatures(json_str):
+                    save_sig.add(ss)
+
     notify('saved signature(s) to {}. Note: signature license is CC0.',
            sigfile_name)
 

--- a/src/sourmash/commands.py
+++ b/src/sourmash/commands.py
@@ -526,7 +526,6 @@ def search(args):
     if args.save_matches:
         notify('saving all matched signatures to "{}"', args.save_matches)
 
-        assert 0
         with SaveSignaturesToLocation(args.save_matches) as save_sig:
             for sr in results:
                 save_sig.add(sr.match)

--- a/tests/test_sourmash.py
+++ b/tests/test_sourmash.py
@@ -1944,10 +1944,7 @@ def test_search_metagenome_downsample_containment():
 def test_search_metagenome_downsample_index(c):
     # does same search as search_metagenome_downsample_containment but
     # rescales during indexing
-    #
-    # for now, this test should fail; we need to clean up some internal
-    # stuff before we can properly implement this!
-    #
+
     testdata_glob = utils.get_test_data('gather/GCF*.sig')
     testdata_sigs = glob.glob(testdata_glob)
 
@@ -1968,6 +1965,38 @@ def test_search_metagenome_downsample_index(c):
     assert ' 29.7%       NC_003197.2 Salmonella enterica subsp. enterica serovar T...' in str(
         c)
     assert '12 matches; showing first 3:' in str(c)
+
+
+def test_search_metagenome_downsample_save_matches(runtmp):
+    c = runtmp
+
+    # does same search as search_metagenome_downsample_containment but
+    # rescales during indexing
+
+    testdata_glob = utils.get_test_data('gather/GCF*.sig')
+    testdata_sigs = glob.glob(testdata_glob)
+
+    query_sig = utils.get_test_data('gather/combined.sig')
+
+    output_matches = runtmp.output('out.zip')
+
+    # downscale during indexing, rather than during search.
+    c.run_sourmash('index', 'gcf_all', *testdata_sigs, '-k', '21',
+                   '--scaled', '100000')
+
+    assert os.path.exists(c.output('gcf_all.sbt.zip'))
+
+    c.run_sourmash('search', query_sig, 'gcf_all', '-k', '21',
+                   '--containment', '--save-matches', output_matches)
+    print(c)
+
+    # is a zip file
+    with zipfile.ZipFile(output_matches, "r") as zf:
+        assert list(zf.infolist())
+
+    # ...with 12 signatures:
+    saved = list(sourmash.load_file_as_signatures(output_matches))
+    assert len(saved) == 12
 
 
 def test_mash_csv_to_sig():

--- a/tests/test_sourmash_sketch.py
+++ b/tests/test_sourmash_sketch.py
@@ -374,8 +374,8 @@ def test_do_sourmash_sketchdna_output_zipfile():
         assert os.path.exists(outfile)
         assert not out # stdout should be empty
 
-        # @CTB do more testing here once panic is fixed!
-        assert 0
+        sigs = list(sourmash.load_file_as_signatures(outfile))
+        assert len(sigs) == 3
 
 
 def test_do_sourmash_sketchdna_output_stdout_valid():

